flight/PiOS/STM32F4xx/pios_dmashot.c

```diff
@@ -0,0 +1,591 @@
+ * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
+ * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License
+ * for more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program; if not, see <http://www.gnu.org/licenses/>
+ *
+ * Additional note on redistribution: The copyright and license notices above
+ * must be maintained in each individual source file that is a derivative work
+ * of this source file; otherwise redistribution is prohibited.
+ */
+
+#include <pios.h>
+#include "pios_dmashot.h"
+
+#define MAX_TIMERS                              8
+
+// This is to do half-word writes where appropriate. TIM2 and TIM5 are 32-bit, the rest
+// are 16-bit. Can do full word writes to 16bit regardless, but it doubles DMA buffer size
+// unnecessarily.
+union dma_buffer {
+	void *ptr;
+	uint32_t *fw;
+	uint16_t *hw;
+};
+
+// Internal structure for timer and DShot configuration.
+struct servo_timer {
+	uint32_t sysclock;                                                              // Flava Flav! Timer's clock frequency.
+
+	const struct pios_dmashot_timer_cfg *dma;                                       // DMA config
+	const struct pios_tim_channel *servo_channels[4];                               // DMAShot configured channels
+
+	// Tracking used TIM channels to allocate least memory.
+	uint8_t low_channel;                                                            // Lowest registered TIM channel
+	uint8_t high_channel;                                                           // Duh
+
+	uint32_t dshot_freq;                                                            // Stores the desired frequency for
+	                                                                                // timer initialization
+	uint16_t duty_cycle_0;                                                          // Calculated duty cycle for 0-bit
+	uint16_t duty_cycle_1;                                                          // And for 1-bit
+
+	union dma_buffer buffer;                                                        // DMA buffer
+};
+
+// DShot signal is 16-bit. Use a pause before and after to delimit signal and quell the timer CC
+// when the message ends.
+#define DMASHOT_MESSAGE_WIDTH                   16
+#define DMASHOT_MESSAGE_PAUSE                   1
+#define DMASHOT_STM32_BUFFER                    (DMASHOT_MESSAGE_PAUSE + DMASHOT_MESSAGE_WIDTH + DMASHOT_MESSAGE_PAUSE)
+
+// Those values are in percent. These represent the Betaflight/BLHeli_s timing,
+// instead of the spec one. Works with KISS RE 24A though.
+#define DSHOT_DUTY_CYCLE_0                      36
+#define DSHOT_DUTY_CYCLE_1                      74
+
+#define TIMC_TO_INDEX(c)                        ((c)>>2)
+
+const struct pios_dmashot_cfg *dmashot_cfg;
+struct servo_timer **servo_timers;
+
+static inline int PIOS_DMAShot_GetFIFOCadence(int num_chan)
```

Comment:
Yes, I'm more saying.. why are you varying the DMA burst size on the memory port?  It's not related to the dma request / transfer burst length.